### PR TITLE
feat(util): Add internal cookie parsing support

### PR DESCRIPTION
**Pull request by santeyio**
Right now I'm keeping to what the python standard library cookie module considers valid cookies. Removes all of the additional features from the standard library module that we don't need or care about for doing simple cookie parsing. 

I wasn't sure where the best place to put the cookie parser and helper methods -- I ended up putting them in with the utils, but let me know if there's somewhere else that I should put them that makes more sense. 

Fixes (#1107)

### CODE REVIEW

**Comment by santeyio** [comment]
I still need to add a few tests to cover some edge cases that the cookie parsing deals with. Will get to that in the next few days.

**Comment by santeyio** [comment]
I think the only thing that's failing in travis is the stuff fixed in PR #1136

**Comment by davidbgk** [comment]
Hello there, we (@pyrates) are working on a similar feature for [Roll](https://github.com/pyrates/roll) and are still hesitating on how to implement it. There are 3 main challengers as of now:

* using `http.cookies` from the standard lib with [known limitations](http://cookies.readthedocs.io/en/latest/#features)
* using [Cookies](http://cookies.readthedocs.io/en/latest/) but it's not really active to say the least with somehow [critical issues](https://gitlab.com/sashahart/cookies/issues/4) stagnant
* using custom implementations [like did Sanic](https://github.com/channelcat/sanic/blob/master/sanic/cookies.py) and that PR to strip off whatever is possible and speed it up

I'd love to get your feedback on these reflections and why you finally chose the latter. Maybe we can join our efforts and create synergies across micro-frameworks to maintain pico parts loosely coupled :)

File: falcon/request.py [hunk 1]
```diff
@@ -893,19 +894,8 @@ def params(self):
     @property
     def cookies(self):
         if self._cookies is None:
-            # NOTE(tbug): We might want to look into parsing
-            # cookies ourselves. The SimpleCookie is doing a
-            # lot if stuff only required to SEND cookies.
             cookie_header = self.get_header('Cookie', default='')
-            parser = SimpleCookie()
```
Comment (davidbgk) [inline falcon/request.py]: I guess you can remove import of `http_cookies`/`SimpleCookie`

File: falcon/request.py [hunk 2]
```diff
@@ -893,19 +894,8 @@ def params(self):
...
-            for cookie_part in cookie_header.split('; '):
-                try:
-                    parser.load(cookie_part)
-                except http_cookies.CookieError:
-                    pass
-            cookies = {}
-            for morsel in parser.values():
-                cookies[morsel.key] = morsel.value
+            cookies = parse_cookies(cookie_header)
```
Comment (davidbgk) [inline falcon/request.py]: Avoid parsing if cookie_header equals default (`''`)?

File: falcon/request.py (hunk 1 above)
Comment (santeyio) [reply falcon/request.py]: right, I did do that in an earlier commit but it seems to have broken the tox 'hug' env build, wanted to ask @kgriffs about that before removing the `SimpleCookie` import?

**Comment by santeyio** [comment]
@davidbgk I think the main motivation for implementing our own cookie parsing is twofold. 

1) Performance reasons. `SimpleCookie` has a bunch of stuff built in around sending cookies, and we don't need that, so removing extraneous code. 

2) We want to be able to parse cookie headers that may have a malformed cookie without throwing out the rest of the cookies (`SimpleCookie` tosses the whole header if there's an illegal cookie). 

That's my understanding of the reasoning, not sure if @kgriffs would add / modify anything in the above.

**Comment by yohanboniface** [comment]
> I think the main motivation for implementing our own cookie parsing is twofold.

Those are indeed legit reasons, but have you considered implementing this in a third party module?

Seems we are in a situation where every Python web framework has its own cookies parsing, but at the end all have the exact same need.
Seems to me that by joining the effort we may have a better implementation (and maintained in the long term).
Now that may be with a new dedicated package, or that may be using and improving an existing one (such as [Cookies](http://cookies.readthedocs.io/en/latest/)).

**Comment by santeyio** [comment]
Well, I certainly wouldn't be opposed to setting this up as a standalone library and putting it up on pip. I didn't realize how confusing the state of cookie parsing was in python land until I started researching for this milestone. I don't know falcon's attitude towards using third party packages though -- @kgriffs what do you think about this?

**Comment by kgriffs** [comment]
@davidbgk @santeyio It seems we may be working toward similar goals. Perhaps it would make sense after all to join forces on a standalone package that both projects (and potentially others) could consume. I agree with @yohanboniface that it is silly for the community to duplicate its efforts in this space. As long as the implementation is performant (should include performance tests) and commits to maintaining an exhaustive test suite, I wouldn't have any problem using it in Falcon. It could potentially be an optional "light-up" dependency similar to how ``ujson`` is treated in the framework today.

**Comment by davidbgk** [comment]
@kgriffs thanks for your answer, that's indeed our intention! We initiated https://github.com/pyrates/biscuits/ since then if you want to join the party :)

**Comment by kgriffs** [comment]
Excellent, thanks everyone. We'll watch https://github.com/pyrates/biscuits/ and add support for it in Falcon when it's ready.
